[Feat] #2 Create CustomTabBarController

ParkSeongGeun opened the pull request:
# 🔑 연관된 Issue

 - 📌  #2 

<br>

# 📘 작업 유형
 - Custom 탭바 구현

<br>

# 📙 작업 내역
 - `Presentation` 모듈 하위에 디렉토리를 생성하여 파일을 생성하였습니다.
    ```
    ├── Presentation
    │   ├── Extension
    │   │   ├── UIColor+.swift
    │   │   └── UIFont+.swift
    │   └── Interface
    │       ├── BaseViewController.swift
    │       ├── Coordinatable.swift
    │       ├── Coordinator.swift
    │       ├── Navigation
    │       │   └── TabBar
    │       │       └── CustomTabBarController.swift
    │       └── ViewModel.swift
     ```
  - 탭을 전환할 때 이전 `viewController`를 제거한 후, 새 `viewController`를 추가하는 방향으로 구현. 
      - 화면 전환 시 이전 탭의 `NavigationStack`이 모두 사라지도록 구현하였습니다.
      - `Coordinator` 패턴 적용은 처음 `TabBarController`에 담기는 `ViewController`에서 하는 것으로 생각했습니다.
      
   - `SceneDelegate.swift` 에서의 수정은 커스텀 탭바 적용 확인을 위한 테스트입니다.
   - `CustomTabBarController.swift` 내부의 `centerButtonDidTap` 함수도 테스트 용도입니다.
<br>

# 🎥  시연 영상

https://github.com/user-attachments/assets/6d57fadc-f21e-4d93-9717-6a247d3d8646


[각 버튼을 눌렀을 때]
<img width="414" alt="image" src="https://github.com/user-attachments/assets/5720527f-0705-46b3-887f-9e6e351880c6">


<br>



### 📝 특이 사항 (Optional)
 - 버튼 모양 / 동작을 구성하는 `UIButton.Configuration`을 사용해보았습니다. -> [공식문서](https://developer.apple.com/documentation/uikit/uibutton/configuration)
 
 - 피드백을 반영하며 리팩토링 하는 과정이 길었습니다. 이에 대한 기록은 학습 정리를 통해 남길 예정입니다.
 <br>

ParkSeongGeun committed:
[Feat] #2 Create CustomTabBarController

- Presentation/Interface/Navigation/TabBar 디렉토리 내부에 TabBarController 생성

ParkSeongGeun committed:
[Fix] Fix Code Convention

futuremirae commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  
+  // MARK: - Navigate Methods
+  private func updateSelectedViewController() {
+    let previousVC = tabItems[previousIndex].viewController
```
제가 기억이 가물가물한데 VC라고 줄여쓰지 않기로 하지 않았나요?!

ParkSeongGeun replied:
맞습니다. 제가 놓쳤었군요, 수정하겠습니다 :)

futuremirae commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  ) {
+    guard viewControllers.count == 2 else {
+      fatalError("CustomTabBarController requires exactly 2 view controllers")
+    }
```
여기서느 꼭 fatalError를 내야하는지 궁금합니다

ParkSeongGeun replied:
현재 탭바가, 2개의 VC(홈, 통계)를 가지도록 설계가 되어있습니다.
결국 처음 앱이 실행되는 과정에서, 탭바 컨트롤러에 두개의 VC를 넣어주어야 하는데, 휴먼 에러를 잡을 수 있는 용도로 넣었습니다!
- os_log로 작성하는 것을 고민했는데, 이를 위해 os를 import 하는 과정이 불필요하다고 판단하였습니다.

Hansangjin98 replied:
의도하신 디버깅 방식을 적용하고 시장에 배포했을 때, 예상치 않게 예외가 발생해 fatalError에 걸린다면 유저에겐 어떤 피드백을 줄 수 있을까요?

유지보수 및 확장의 과정에서 예외를 발생시킬 수 있는 로직이 추가된 채로 배포가 나가게 된다면 어떻게 대처할 수 있을까요?

ParkSeongGeun replied:
결국 에러가 난 과정에서 앱을 종료하기 위해 fatalError를 썼었는데, 사용자는 단순히 "앱이 꺼졌다" 로 인식만 할 수 있겠군요.

서칭을 해보니
애플이 권하는 앱 종료 방법은
```
1. 유저에게 어떤 문제가 발생했는지와 선택 가능한 액션을 Alert 로 보여준다.
2. 갑자기 앱이 꺼지는 것처럼 느껴지지 않게, 홈 스크린으로 돌아가는 애니메이션과 함께 우아하게 앱을 종료시킨다.
```
와 같이 권장하는 방식이 있네요.

현재, 각자 구현을 하는 과정에서 각각의 에러 상황이 발생하겠지만, 이는 Core와 같이 최상단 모듈에서 구현을 추가적으로 해야할 거 같다고 생각이 드네요.

다만, 현재 상황에서 결국 에러를 핸들링을 어떻게 할 지는 논의가 필요할 거 같습니다 :)

Hansangjin98 replied:
배열의 인덱스에 직접적으로 접근을 하고 있어서 guard 문으로 배열의 count를 체크하고 있는 것 같은데
인덱스에 직접적으로 접근하는 방식 대신 순회하는 방식으로 만들어서 보여주고 디버깅을 위한 Log를 출력하는 건 어떠신가요~?

futuremirae commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  }
+  
+  // MARK: - UI Components
+  private lazy var tabBarView: UIView = {
```
lazy를 넣은 이유가 ViewController가 생성된 이후에 초기화 될 수 있도록 하기 위해서이신건가요?
self를 사용하게 될때는 lazy를 꼭 붙여줘야하는데, 
여기서 붙여주신 이유가 궁급합니다!

ParkSeongGeun replied:
미래님이 말씀하신 이유도 근거 중 하나입니다.
추가적으로 현재 cornerRadius, size 등을 Constants enum에서 사용 중인데, 
인스턴스 초기화 시점에 접근을 못하는 경우를 방지 위하여 lazy를 사용했습니다.
- 기존 스프린트 때부터 UI요소를 사용할 때 lazy를 대부분 붙였던 것 같습니다.

Hansangjin98 replied:
1) tabBarView는 Constants의 static 상수에 접근하고 있네요. 초기화 시점에 해당 상수에 접근하지 못하는 경우는 어떤 경우가 있을까요?

2) 기존 스프린트 때부터 UI요소를 사용할 때 lazy를 대부분 붙였던 이유가 궁금합니다!

3) Apple은 변수 초기화 과정을 왜 default로 lazy하게 두지 않았을까요? lazy하게 두었을 때 발생할 수 있는 단점이 있었을까요?

ParkSeongGeun replied:
미래님 / 상진님이 언급해주셔서 다시금 lazy var, let으로 선언하는 것의 차이에 대해서 조금 찾아봤습니다.

1. 현재 버튼을 제외한 필수적인 UIComponent들이 선언되는 과정에서는 lazy var 로 선언할 이유는 없어보입니다.

2. 현재 코드에서의 일관성 유지라고 (개인적으로) 생각하여 모두 lazy var를 적용을 했었습니다.

3. 조사를 해보니 아래와 같은 대표적인 단점이 있군요. -> [출처](https://www.quora.com/Why-we-not-set-all-variable-to-lazy-in-Swift)
```
컴파일러가 생성한 추가 코드는 실제로 CPU가 처리해야 하는 작업입니다
불필요한 lazy 사용은:
    - 프로그램 속도를 저하시킴
    - 앱 크기를 증가시킴
    - CPU에 추가 작업을 부과함
결론: lazy는 정말 필요한 경우에만 사용해야 하는 기능이며, 무분별한 사용은 앱의 성능과 크기에 부정적인 영향을 미칠 수 있습니다.
```

* 결론적으로 필요한 부분에서만 적용할 수 있도록 해야겠네요 ;)

clxxrlove replied:
저도 스프린트 때, 사용하기 편하다는 이유로 `lazy var`를 자주 사용했는데요, 관련 코멘트가 많아 저도 어떤 문제가 있을지 알아봤습니다. 메모리 문제나 흐름 관련해서 주의해야 할 부분들이 꽤나 있는거 같네요. 

```
1. 메모리 관리 문제 (retain cycle)
  - lazy var는 초기화 시점에 self에 접근하므로, 만약 lazy var의 초기화 과정에서 self의 특정 요소에 강하게 의존하면 **순환 참조(retain cycle)**가 발생할 수 있습니다. 이 경우 메모리 누수로 이어질 수 있습니다.
  - 예를 들어, lazy var 프로퍼티가 클로저 내에서 self를 강하게 캡처하는 경우 문제가 됩니다.
2. 초기화 순서 문제
  - lazy 프로퍼티는 첫 사용 시점에 초기화됩니다. 따라서 viewDidLoad 전에 접근하는 경우에는 초기화되지 않아 예상과 다른 동작을 할 수 있습니다. 이는 lazy var에 의존하는 뷰 구성에서 예상치 못한 결과를 낳을 수 있습니다.
```

Hansangjin98 replied:
추가적으로 lazy 타입 변수는 thread-safe 하지 않기 때문에 활용할 때도 thread 문제가 발생하지 않도록 처리가 필요합니다!

futuremirae commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  
+  private func setupTabBar() {
+    view.addSubview(tabBarView)
+    tabBarView.snp.makeConstraints { make in
```
이 부분 $0이렇게도 사용할수 있던데 이부분도 같이 통일할지 이야기를 나눠보면 좋을 것 같아요!

Hansangjin98 commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+    button.setImage(
+      UIImage(systemName: "music.mic", withConfiguration: UIImage.SymbolConfiguration(pointSize: Constants.centerButtonIconSize, weight: .medium)),
+      for: .normal
+    )
```
UIImage를 생성할 때 파일 이름에서 휴먼 에러가 발생하지 않게, 또는 이미지가 변경 되었을 때 유지보수가 편하도록 별도로 관리하면 좋겠네요!

이 부분은 제가 작업해서 반영하도록 하겠습니다!

Hansangjin98 commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  }
+  
+  // MARK: - Navigate Methods
+  private func updateSelectedViewController() {
+    let previousVC = tabItems[previousIndex].viewController
+    previousVC.willMove(toParent: nil)
+    previousVC.view.removeFromSuperview()
+    previousVC.removeFromParent()
+    
+    let currentVC = tabItems[currentIndex].viewController
+    addChild(currentVC)
+    view.insertSubview(currentVC.view, belowSubview: tabBarView)
+    
+    currentVC.view.snp.makeConstraints { make in
+      make.edges.equalToSuperview()
+    }
+    
+    currentVC.didMove(toParent: self)
+    updateButtonAppearance()
+  }
```
탭을 누를 때 마다 이전 탭의 화면 스택이 모두 날라가게 될 것 같은데 의도하신 흐름일까요?

ParkSeongGeun replied:
네 맞습니다. 

세부 화면으로 진입하여도 다른 탭을 클릭하게 되면 이전 화면 스택이 날라가는 것으로 인지하였습니다.
- 현재는 Coordinator 패턴이 적용되지 않는 화면 흐름으로 진행하고 있어, 위와 같이 설정하였습니다.

Hansangjin98 commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+    }
+  }
+  
+  private func setupTabButtons() {
+    let buttonWidth = view.bounds.width / 3
+    
+    tabItems.enumerated().forEach { index, item in
+      let button = createTabButton(for: item, at: index)
+      tabBarView.addSubview(button)
+      
+      button.snp.makeConstraints { make in
+        make.bottom.equalToSuperview().offset(Constants.buttonBottomOffset)
+        make.height.equalTo(Constants.buttonHeight)
+        make.width.equalTo(buttonWidth)
+        
+        if index == 0 {
+          make.leading.equalToSuperview()
+        } else {
+          make.trailing.equalToSuperview()
+        }
+      }
+      
+      tabButtons.append(button)
+    }
+    
+    updateButtonAppearance()
+  }
```
탭 바 버튼이 추가되거나 줄어든다면 어떻게 대응할 수 있을까요?
코드의 변경이 발생해야 할까요?

ParkSeongGeun replied:
`let buttonWidth = view.bounds.width / CGFloat(tabItems.count + 1)` 와 같이 변경하였습니다.

Hansangjin98 replied:
```swift
        if index == 0 {
          make.leading.equalToSuperview()
        } else {
          make.trailing.equalToSuperview()
        }
```

이 부분도 처리가 필요하지 않을까요?

ParkSeongGeun replied:
맞습니다. 코멘트를 못 남겼는데, 해당 부분도 반영할 예정입니다 :)
- 이와 같은 부분에서도 느끼게 되지만, 유지보수 관점에서 수정이 필요한 코드인 것 같습니다.

Hansangjin98 commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  }
+}
+
+extension CustomTabBarController {
+  private enum Constants {
+    static let tabBarHeight: CGFloat = 90
+    static let tabBarCornerRadius: CGFloat = 20
+    static let centerButtonSize: CGFloat = 80
+    static let centerButtonIconSize: CGFloat = 30
+    static let buttonIconSize: CGFloat = 18
+    static let buttonTitleSize: CGFloat = 10
+    static let buttonSpacing: CGFloat = 5
+    static let buttonBottomOffset: CGFloat = -20
+    static let buttonHeight: CGFloat = 50
+    static let inactiveAlpha: CGFloat = 0.5
+  }
+}
```
Constants만 별도의 extension으로 분리하신 이유가 궁금합니다!

ParkSeongGeun replied:
휴먼 에러 방지를 위해 Constants를 설정하였고, 가독성 향상을 위해 따로 extension으로 분리하였습니다.

Hansangjin98 commented on Heim/Presentation/Presentation/Interface/Navigation/TabBar/CustomTabBarController.swift:
```diff
@@ -0,0 +1,235 @@
+  }
+  
+  // MARK: - Public Methods
+  public func setViewControllers(
+    _ viewControllers: [UIViewController]
+  ) {
```
탭의 화면에 대해 Coordinator가 아니라 ViewController를 직접 주입받고 있는데 추후 수정되는 부분일까요?

ParkSeongGeun replied:
위에서도 말씀드렸다싶이, UI / 화면 이동 기능만을 위한 탭바 구성을 위해 현재 Coordinator를 주입받지 않고, VC를 직접 주입받도록 하였습니다.
- 이를 토대로 SceneDelegate.swift에서 테스트를 하였습니다.
- 각 Feature에 맞는 VC, VM이 나오게 된다면 Coordinator를 주입받도록 수정할 예정입니다.

Hansangjin98 reviewed (commented):
설계에 대해 다시 한번 고민해 주시면 좋을 것 같아요😀

TabBar의 흐름을 ViewController로 구현함으로써 TabBar의 UI와 Action에 대해 모두 ViewController가 담당하고 있는데, 현재의 코드는 이후 변화에 대응하기 어려운 상태라는 개인적인 생각이 들어요.
ViewController가 명백히 상태와 행위를 지니고 있기 때문에 유지보수에 용이하게 개선할 수 있을 것 같습니다!
저희가 적용하고 있는 디자인 패턴과 'TabBar의 Action에 대해 Unit Test를 어떻게 진행할 수 있을지'를 함께 생각해보면 한결 수월하실 것 같습니다.

개인적으로, UITabBarController를 사용하지 않고 ViewController로 직접 구현한 이유도 궁금합니다!

ParkSeongGeun committed:
[Refactor] #2 PR Review

- 기존 디렉토리에서 Presentation/TabBar로 이동
    - 추후 develop에서 merge되면 Presentation/Common/TabBar로 옮길 예정

- ViewController, ViewModel로 분리
    - ViewModel: Presentation 로직을 담당
    - ViewController: UI 컴포넌트 초기화 및 설정, ViewModel과의 바인딩 설정

ParkSeongGeun commented on Heim/Presentation/Presentation/TabBar/CustomTabBarViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        make.height.equalTo(Constants.buttonHeight)
+        make.width.equalTo(buttonWidth)
+        
+        if index == 0 {
```
'첫번째', '마지막', '그 외' 3가지로 분류하여 레이아웃을 잡아주려고 했습니다.
다만, 현재는 Tap을 할 화면이 2가지 밖에 없어, 두었습니다.

ParkSeongGeun reviewed (commented):
[수정내역]
- 기존 디렉토리에서 Presentation/TabBar로 이동
    - 추후 develop에서 merge되면 Presentation/Common/TabBar로 옮길 예정

- ViewController, ViewModel로 분리
    - ViewModel: Presentation 로직을 담당
    - ViewController: UI 컴포넌트 초기화 및 설정, ViewModel과의 바인딩 설정

ParkSeongGeun committed:
[Refactor] #2 PR Review

- View, ViewController, ViewModel 간의 책임 분리 / 역할 재정립
